File: Examples/app/src/main/java/com/scichart/examples/fragments/showcase/audioAnalyzer/StubAudioAnalyzerDataProvider.java
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2020. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// StubAudioAnalyzerDataProvider.java is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

package com.scichart.examples.fragments.showcase.audioAnalyzer;

import com.scichart.examples.fragments.base.DataProviderBase;

import java.util.Random;
import java.util.concurrent.TimeUnit;

public class StubAudioAnalyzerDataProvider extends DataProviderBase<AudioData> implements IAudioAnalyzerDataProvider {
    private final int bufferSizeInShorts = 2048;
    private final AudioData audioData = new AudioData(bufferSizeInShorts);

    private final IYValuesProvider provider = new AggregateYValueProvider(
      new FrequencySinewaveYValueProvider(8000, 0, 0, 1, 0.0000005),
      new NoisySinewaveYValueProvider(8000, 0, 0.000032, 200),
      new NoisySinewaveYValueProvider(6000, 0, 0.000016, 100),
      new NoisySinewaveYValueProvider(4000, 0, 0.000064, 100)
    );

    private long time = 0L;

    public StubAudioAnalyzerDataProvider() {
        super(20L, TimeUnit.MILLISECONDS);
    }

    @Override
    protected AudioData onNext() {
        final long[] xItems = audioData.xData.getItemsArray();
        final short[] yItems = audioData.yData.getItemsArray();

        for (int i = 0; i < bufferSizeInShorts; i++) {
            xItems[i] = time++;
            yItems[i] = provider.getYValueForIndex(time);
        }

        return audioData;
    }

    @Override
    public int getBufferSize() {
        return bufferSizeInShorts;
    }

    @Override
    public int getSampleRate() {
        return 44100;
    }

    private interface IYValuesProvider {
        short getYValueForIndex(long index);
    }

    private static class NoisySinewaveYValueProvider implements IYValuesProvider {
       private final Random random = new Random();

       private final double amplitude;
       private final double phase;
       private final double noiseAmplitude;
       private final double wn;

        public NoisySinewaveYValueProvider(double amplitude, double phase, double frequency, double noiseAmplitude) {
            this.amplitude = amplitude;
            this.phase = phase;
            this.noiseAmplitude = noiseAmplitude;
            this.wn = 2 * Math.PI * frequency;
        }

        @Override
        public short getYValueForIndex(long index) {
            return (short)(amplitude * Math.sin(index * wn + phase) + (random.nextDouble() - .5) * noiseAmplitude);
        }
    }

    private static class FrequencySinewaveYValueProvider implements IYValuesProvider {
        private final double amplitude;
        private final double phase;

        private final double minFrequency;
        private final double maxFrequency;
        private final double step;

        private double frequency;

        public FrequencySinewaveYValueProvider(double amplitude, double phase, double minFrequency, double maxFrequency, double step) {
            this.amplitude = amplitude;
            this.phase = phase;

            this.minFrequency = minFrequency;
            this.maxFrequency = maxFrequency;
            this.step = step;

            this.frequency = minFrequency;
        }

        @Override
        public short getYValueForIndex(long index) {
            this.frequency = frequency <= maxFrequency ? frequency + step : minFrequency;

            final double wn = 2 * Math.PI * frequency;
            return (short) (amplitude * Math.sin(index * wn + phase));
        }
    }

    private static class AggregateYValueProvider implements IYValuesProvider {
        private final IYValuesProvider[] providers;

        private AggregateYValueProvider(IYValuesProvider... providers) {
            this.providers = providers;
        }

        @Override
        public short getYValueForIndex(long index) {
            double sum = 0;
            for (int i = 0; i < providers.length; i++) {
                sum += providers[i].getYValueForIndex(index);
            }
            return (short) sum;
        }
    }
}

File: Showcase/app/src/main/kotlin/com/scichart/scishowcase/model/audioAnalyzer/StubAudioAnalyzerDataProvider.kt
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2017. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// StubAudioAnalyzerDataProvider.kt is part of the SCICHART® Showcases. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

package com.scichart.scishowcase.model.audioAnalyzer

import com.scichart.scishowcase.model.DataProviderBase
import java.util.*
import java.util.concurrent.TimeUnit

class StubAudioAnalyzerDataProvider(private val bufferSizeInShorts: Int = 2048) : DataProviderBase<AudioData>(20L, TimeUnit.MILLISECONDS), IAudioAnalyzerDataProvider {

    var time = 0L

    val audioData = AudioData(bufferSizeInShorts)

    private val provider = AggregateYValueProvider(arrayOf(
            FrequencySinewaveYValueProvider(8000.0, 0.0, 0.0, 1.0, 0.0000005),
            NoisySinewaveYValueProvider(8000.0, 0.0, 0.000032, 200.0),
            NoisySinewaveYValueProvider(6000.0, 0.0, 0.000016, 100.0),
            NoisySinewaveYValueProvider(4000.0, 0.0, 0.000064, 100.0)
    ))

    override fun onNext(): AudioData {
        val xItemsArray = audioData.xData.itemsArray
        val yItemsArray = audioData.yData.itemsArray

        for (index in 0 until bufferSizeInShorts) {
            xItemsArray[index] = time++
            yItemsArray[index] = provider.getYValueForIndex(time)
        }

        return audioData
    }

    override fun getBufferSize(): Int = bufferSizeInShorts

    override fun getSampleRate(): Int = 44100

    private interface IYValueProvider {
        fun getYValueForIndex(index: Long): Short
    }

    private class NoisySinewaveYValueProvider(private val amplitude: Double, private val phase: Double, private val freq: Double, private val noiseAmplitude: Double) : IYValueProvider {
        private val random = Random()
        override fun getYValueForIndex(index: Long): Short {
            val wn = 2 * Math.PI * freq
            return (amplitude * Math.sin(index * wn + phase) + (random.nextDouble() - 0.5) * noiseAmplitude).toShort()
        }
    }

    private class FrequencySinewaveYValueProvider(private val amplitude: Double, private val phase: Double, private val minFrequency: Double, private val maxFrequency: Double, private val step: Double) : IYValueProvider {
        private var frequency = minFrequency

        override fun getYValueForIndex(index: Long): Short {
            frequency = if(frequency <= maxFrequency) frequency + step else minFrequency

            val wn = 2 * Math.PI * frequency
            return (amplitude * Math.sin(index * wn + phase)).toShort()
        }
    }

    private class AggregateYValueProvider(val providers: Array<IYValueProvider>) : IYValueProvider {
        override fun getYValueForIndex(index: Long): Short {
            return providers.sumByDouble { it.getYValueForIndex(index).toDouble() }.toShort()
        }

    }
}